cli/pkg/kctrl/cmd/package/release/helm_openapi_schema_gen.go

```diff
@@ -0,0 +1,373 @@
+			OpenAPIv3: runtime.RawExtension{Raw: nil},
+		}, nil
+	}
+	if document.Kind != yaml3.DocumentNode {
+		// return proper error message
+		return nil, fmt.Errorf("invalid node kind supplied: %d", document.Kind)
+	}
+	if document.Content[0].Kind != yaml3.MappingNode {
+		return nil, fmt.Errorf("values file must resolve to a map (was %d)", document.Content[0].Kind)
+	}
+	openAPIProperties, err := h.calculateProperties(nil, document.Content[0])
+	if err != nil {
+		return nil, err
+	}
+
+	bytes, err := yaml2.Marshal(h.convertToYAML(openAPIProperties))
+	if err != nil {
+		return nil, err
+	}
+	jsonEncodedBytes, err := yaml.YAMLToJSON(bytes)
+	if err != nil {
+		return nil, err
+	}
+	return &kcdatav1alpha1.ValuesSchema{
+		OpenAPIv3: runtime.RawExtension{Raw: jsonEncodedBytes},
+	}, nil
+}
+
+func (h HelmValuesSchemaGen) readValuesFile() ([]byte, error) {
+	helmChartDir := h.dirPath
+	fileInfo, err := os.Stat(helmChartDir)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read %s: %s", h.dirPath, err.Error())
+	}
+	if !fileInfo.IsDir() {
+		return nil, fmt.Errorf("expected %s to be directory", h.dirPath)
+	}
+
+	_, err = os.Stat(filepath.Join(helmChartDir, "values.yaml"))
+	if err != nil {
+		// It is possible that helm chart doesn't have values.yml file.
+		if os.IsNotExist(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	return os.ReadFile(filepath.Join(helmChartDir, "values.yaml"))
+}
+
+func (h HelmValuesSchemaGen) fileExists(filename string) bool {
+	info, err := os.Stat(filename)
+	if os.IsNotExist(err) {
+		return false
+	}
+	return !info.IsDir()
+}
+func (h HelmValuesSchemaGen) convertToYAML(val interface{}) interface{} {
+	switch typedVal := val.(type) {
+	case *Map:
+		result := yaml2.MapSlice{}
+		typedVal.Iterate(func(k, v interface{}) {
+			result = append(result, yaml2.MapItem{
+				Key:   k,
+				Value: h.convertToYAML(v),
+			})
+		})
+		return result
+	default:
+		return val
+	}
+}
+
+func (h HelmValuesSchemaGen) calculateProperties(key *yaml3.Node, value *yaml3.Node) (*Map, error) {
+	switch value.Kind {
+	case yaml3.MappingNode:
+		var apiKeys openAPIKeys
+		description, isPresent := h.getDescriptionFromNode(key)
+		if isPresent {
+			apiKeys = append(apiKeys, &MapItem{Key: descriptionProp, Value: description})
+		}
+		apiKeys = append(apiKeys, &MapItem{Key: typeProp, Value: objectVal})
+
+		if len(value.Content) == 0 {
+			apiKeys = append(apiKeys, &MapItem{Key: defaultProp, Value: "{}"})
+		} else {
+			var properties openAPIKeys
+			for i := 0; i < len(value.Content); i += 2 {
+				k := value.Content[i]
+				v := value.Content[i+1]
+				calculatedProperties, err := h.calculateProperties(k, v)
+				if err != nil {
+					return nil, err
+				}
+				valueItems := calculatedProperties.Items[0]
+				properties = append(properties, &MapItem{Key: valueItems.Key, Value: valueItems.Value})
+			}
+			apiKeys = append(apiKeys, &MapItem{Key: propertiesProp, Value: &Map{Items: properties}})
+		}
```

Comment:
```suggestion
                var properties openAPIKeys
		for i := 0; i < len(value.Content); i += 2 {
			k := value.Content[i]
			v := value.Content[i+1]
			calculatedProperties, err := h.calculateProperties(k, v)
			if err != nil {
				return nil, err
			}
			valueItems := calculatedProperties.Items[0]
			properties = append(properties, &MapItem{Key: valueItems.Key, Value: valueItems.Value})
		}
		if len(value.Content) == 0 {
			apiKeys = append(apiKeys, &MapItem{Key: defaultProp, Value: "{}"})
		} else {
			apiKeys = append(apiKeys, &MapItem{Key: propertiesProp, Value: &Map{Items: properties}})
		}
```

Reply:
I prefer it like above, but If you still think we should do like below, I am open to do that.

Reply:
With the second approach we are relying on the condition in the for loop itself to keep the normal code path at minimal indentation (which is in general more readable)

Reply:
Adopted